### PR TITLE
Isolate source AppHost deployment state

### PR DESCRIPTION
**Pull request by spboyer**
## Description

Source-file AppHosts in the same directory currently share a directory-scoped deployment-state identity. That allows one AppHost to enumerate, update, or clear state that belongs to a sibling AppHost.

This change gives source-file AppHosts a source-specific deployment-state identity while preserving safe access to legacy state during migration. Legacy storage remains immutable because its ownership is ambiguous. Each source AppHost records an atomic migration snapshot, canonical deltas, and tombstones in its own sidecar so interrupted writes cannot revive deleted state. Parent-section saves use three-way leaf merging, preserving independent concurrent changes and preventing unchanged sibling descendants from being adopted.

The deployment-state API also gains `AcquireCurrentSectionAsync`, which lets deployment targets enumerate only canonical state while ordinary reads retain migration fallback. This is a prerequisite for the Azure Container Apps Sandboxes work in #19008.

### User-facing usage

Deployment-state consumers that must exclude legacy-only entries can acquire the current AppHost's canonical section:

```csharp
var section = await deploymentStateManager.AcquireCurrentSectionAsync(
    "Azure:Sandboxes",
    cancellationToken);
```

Validation includes focused coverage for source identity, partial and repeated migration, broad parent saves, concurrent writers, tombstones, clear behavior, malformed metadata, interrupted compatibility writes, and early configuration loading.

Fixes # (issue)

## Checklist

- Is this feature complete?
  - [x] Yes. Ready to ship.
  - [ ] No. Follow-up changes expected.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [x] Yes
  - [ ] No
- Did you add public API?
  - [x] Yes
    - If yes, did you have an API Review for it?
      - [ ] Yes
      - [x] No
    - Did you add `<remarks />` and `<code />` elements on your triple slash comments?
      - [x] Yes
      - [ ] No
  - [ ] No
- Does the change make any security assumptions or guarantees?
  - [ ] Yes
    - If yes, have you done a threat model and had a security review?
      - [ ] Yes
      - [ ] No
  - [x] No

### CODE REVIEW

File: src/Aspire.Hosting/Pipelines/IDeploymentStateManager.cs [hunk 1]
```diff
@@ -25,6 +25,15 @@ public interface IDeploymentStateManager
...
     Task<DeploymentStateSection> AcquireSectionAsync(string sectionName, CancellationToken cancellationToken = default);
 
+    /// <summary>
+    /// Acquires a specific section from the current deployment state without falling back to legacy state.
+    /// </summary>
+    /// <param name="sectionName">The name of the section to acquire (e.g., "Parameters", "Azure").</param>
+    /// <param name="cancellationToken">The cancellation token.</param>
+    /// <returns>A StateSection containing the current section data and version information.</returns>
+    Task<DeploymentStateSection> AcquireCurrentSectionAsync(string sectionName, CancellationToken cancellationToken = default) =>
+        AcquireSectionAsync(sectionName, cancellationToken);
```
[the remark this answers is not in the document]
Comment (spboyer) [reply src/Aspire.Hosting/Pipelines/IDeploymentStateManager.cs]: Fixed in 7065f0839. `AcquireCurrentSectionAsync` is now a required interface operation, and every implementation explicitly supplies current-only semantics rather than inheriting legacy-inclusive behavior.

File: src/Aspire.Hosting/Pipelines/Internal/FileDeploymentStateManager.cs [hunk 2]
```diff
@@ -96,20 +232,46 @@ protected override async Task SaveStateToStorageAsync(JsonObject state, Cancella
...
-            DirectoryHelper.CreateWithOwnerOnlyPermissions(deploymentStateDirectory);
-            await File.WriteAllTextAsync(
-                deploymentStatePath,
-                flattenedSecrets.ToJsonString(UserSecretsJsonOptions.s_instance),
-                cancellationToken).ConfigureAwait(false);
+            using (await FileLock.AcquireAsync($"{deploymentStatePath}.lock", cancellationToken).ConfigureAwait(false))
+            {
+                if (sectionName is null)
+                {
+                    _currentState = state.DeepClone().AsObject();
```
[the remark this answers is not in the document]
Comment (spboyer) [reply src/Aspire.Hosting/Pipelines/Internal/FileDeploymentStateManager.cs]: Fixed in 7065f0839. Full-state replacement now clears claims, legacy snapshots, and cached legacy state, and disables fallback before persisting migration metadata. The regression also verifies a subsequent section save through the same manager cannot resurrect a legacy value.

File: src/Aspire.Hosting/DistributedApplicationBuilder.cs [hunk 3]
```diff
@@ -1000,26 +1024,35 @@ private void LoadDeploymentState(string appHostSha)
...
+                deploymentStatePath,
+                legacyDeploymentStatePath).GetAwaiter().GetResult();
+            if (effectiveState.Count > 0)
+            {
+                var flattenedState = JsonFlattener.FlattenJsonObject(effectiveState);
+                using var stream = new MemoryStream(Encoding.UTF8.GetBytes(flattenedState.ToJsonString()));
+                _innerBuilder.Configuration.AddJsonStream(stream);
+            }
         }
         catch { }
```
[the remark this answers is not in the document]
Comment (spboyer) [reply src/Aspire.Hosting/DistributedApplicationBuilder.cs]: Fixed in 7065f0839. Builder construction now uses a synchronous effective-state loader under the same cross-process lock, with targeted handling and debug diagnostics for I/O, access, JSON/data, configuration-format, and lock-timeout failures. Invalid root shapes and case-insensitive duplicate configuration keys are covered.

File: tests/Aspire.Hosting.Tests/Publishing/DeploymentStateManagerTests.cs [hunk 4]
```diff
@@ -418,14 +434,366 @@ public void GetStatePath_UsesDefaultAspireHome_WhenAspireHomeIsNotConfigured()
...
+        Assert.NotEqual(legacyPath, currentPath);
+        Assert.True(File.Exists(currentPath));
+        Assert.True(File.Exists(legacyPath));
+
+        var currentStateManager = CreateFileDeploymentStateManager(currentSha);
+        var currentSection = await currentStateManager.AcquireSectionAsync("Migration");
+        Assert.True(currentSection.Data["legacy"]?.GetValue<bool>());
+        Assert.True(currentSection.Data["current"]?.GetValue<bool>());
+        await migratingStateManager.ClearAllStateAsync();
+        await legacyStateManager.ClearAllStateAsync();
```
[the remark this answers is not in the document]
Comment (spboyer) [reply tests/Aspire.Hosting.Tests/Publishing/DeploymentStateManagerTests.cs]: Fixed in 7065f0839. The migration test now performs both state cleanups from `finally`, so assertion failures cannot leave shared deployment state behind.

File: src/Aspire.Hosting/Pipelines/Internal/FileDeploymentStateManager.cs [hunk 5]
```diff
@@ -96,20 +271,50 @@ protected override async Task SaveStateToStorageAsync(JsonObject state, Cancella
...
+                    if (TryGetNestedPropertyValue(_legacyStateSnapshot, sectionName) is null &&
+                        legacySectionData is not null)
+                    {
+                        SetNestedPropertyValue(_legacyStateSnapshot, sectionName, legacySectionData.DeepClone().AsObject());
+                    }
+                    SetNestedPropertyValue(_currentState, sectionName, null);
+                    _claimedSectionNames.RemoveWhere(name =>
+                        string.Equals(name, sectionName, StringComparison.Ordinal) ||
+                        name.StartsWith($"{sectionName}:", StringComparison.Ordinal));
+                    ApplyStateDelta(_currentState, sectionName, legacySectionData, desiredSectionData, _claimedSectionNames);
```
[the remark this answers is not in the document]
Comment (spboyer) [reply src/Aspire.Hosting/Pipelines/Internal/FileDeploymentStateManager.cs:365]: Fixed in 3472c81f5. Migration now preserves the legacy section as a `JsonNode`, snapshots scalar values with the node-aware helper, and records a claim when the scalar is deleted. A restart regression covers a legacy `Parameters:secret` string deletion.

File: src/Aspire.Hosting/DistributedApplicationBuilder.cs [hunk 6]
```diff
@@ -1000,26 +1025,38 @@ private void LoadDeploymentState(string appHostSha)
...
     }
 
+    private static string GetDeploymentStatePath(string appHostSha, string environment) =>
+        Path.Combine(
+            System.Environment.GetFolderPath(System.Environment.SpecialFolder.UserProfile),
+            ".aspire",
+            "deployments",
+            appHostSha,
+            $"{environment}.json"
+        );
```
[the remark this answers is not in the document]
Comment (spboyer) [reply src/Aspire.Hosting/DistributedApplicationBuilder.cs]: Fixed in 3472c81f5. Builder-time loading now uses the same `FileDeploymentStateManager.GetStatePath` resolver as runtime state operations, so configured `ASPIRE_HOME` and the default fallback cannot diverge.

File: src/Aspire.Hosting/Pipelines/Internal/FileDeploymentStateManager.cs [hunk 7]
```diff
@@ -119,4 +324,350 @@ await File.WriteAllTextAsync(
...
+        JsonObject currentState,
+        IEnumerable<string> claimedSectionNames)
+    {
+        foreach (var sectionName in claimedSectionNames)
+        {
+            if (!NestedPropertyExists(currentState, sectionName))
+            {
+                SetNestedNodeValue(effectiveState, sectionName, value: null, valueExists: false);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (spboyer) [reply src/Aspire.Hosting/Pipelines/Internal/FileDeploymentStateManager.cs:639]: Fixed in b89abec56. Claimed paths now make the canonical node authoritative whether it is present or absent: present canonical nodes replace the already-merged effective subtree, while absent nodes remove it. Added a restart regression that deletes the claimed `Azure` parent, recreates `Azure:Sandboxes:new`, and verifies legacy parent properties and sibling sandboxes do not reappear. The focused state/builder suite passes all 112 tests.

File: src/Aspire.Hosting/Pipelines/Internal/FileDeploymentStateManager.cs [hunk 8]
```diff
@@ -86,7 +200,68 @@ internal static bool IsValidEnvironmentName(string environmentName)
...
+
+        var mergedSection = base.GetSectionState(state, sectionName, includeLegacyState)?.DeepClone();
+        foreach (var claimedSectionName in _claimedSectionNames.Where(name => name.StartsWith($"{sectionName}:", StringComparison.Ordinal)))
+        {
+            if (!NestedPropertyExists(_currentState, claimedSectionName) &&
+                mergedSection is JsonObject mergedObject)
+            {
+                SetNestedPropertyValue(mergedObject, claimedSectionName[(sectionName.Length + 1)..], null);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (spboyer) [reply src/Aspire.Hosting/Pipelines/Internal/FileDeploymentStateManager.cs:301]: Fixed in 654065338. Parent-section reads now replace every claimed descendant with the authoritative canonical subtree when present and remove it when absent, using the same explicit-null/absence semantics as effective-state construction. Added a restart regression that deletes `Azure:Sandboxes:first`, recreates `first:new`, and verifies a parent `Azure:Sandboxes` read cannot resurrect legacy properties while retaining unclaimed siblings. The focused state/builder suite passes all 114 tests.

File: src/Aspire.Hosting/Pipelines/Internal/DeploymentStateManagerBase.cs [hunk 9]
```diff
@@ -138,7 +189,7 @@ public async Task<DeploymentStateSection> AcquireSectionAsync(string sectionName
             JsonObject? data = null;
             string? value = null;
 
-            var sectionData = TryGetNestedPropertyValue(_state, sectionName);
+            var sectionData = GetSectionState(_state, sectionName, includeLegacyState);
```
[the remark this answers is not in the document]
Comment (spboyer) [reply src/Aspire.Hosting/Pipelines/Internal/DeploymentStateManagerBase.cs:192]: Fixed in 654065338. Scalar acquisition now snapshots `OriginalData` after populating the root scalar value, so an unchanged stale save participates correctly in the three-way merge and preserves a concurrent update. Added `UnchangedScalarSavePreservesConcurrentUpdate`; the focused state/builder suite passes all 114 tests.

File: src/Aspire.Hosting/Pipelines/Internal/FileDeploymentStateManager.cs [hunk 10]
```diff
@@ -96,20 +278,50 @@ protected override async Task SaveStateToStorageAsync(JsonObject state, Cancella
...
+                    // Reload while holding the cross-process lock so repeated or concurrent
+                    // process starts cannot overwrite sections saved by another instance.
+                    _currentState = await LoadStateFileAsync(deploymentStatePath, cancellationToken).ConfigureAwait(false);
+                    await LoadMigrationStateAsync(deploymentStatePath, cancellationToken).ConfigureAwait(false);
+                    var sectionData = TryGetNestedPropertyValue(state, sectionName) as JsonObject;
+                    var legacySectionData = TryGetNestedPropertyValue(GetLegacyFallbackState(), sectionName);
+                    var latestEffectiveState = MergeState(GetLegacyFallbackState(), _currentState);
+                    ApplyClaimedSections(latestEffectiveState, _currentState, _claimedSectionNames);
+                    var latestSectionData = TryGetNestedPropertyValue(latestEffectiveState, sectionName);
+                    var desiredSectionData = ApplyChanges(latestSectionData, originalSectionData, sectionData) as JsonObject;
```
[the remark this answers is not in the document]
Comment (spboyer) [reply src/Aspire.Hosting/Pipelines/Internal/FileDeploymentStateManager.cs]: Fixed in 168629d95. Canonical scalar `JsonValue` nodes are now normalized to the section scalar wrapper before the three-way merge, with a regression that seeds canonical scalar state without a migration sidecar.

File: src/Aspire.Hosting/Pipelines/Internal/FileDeploymentStateManager.cs [hunk 11]
```diff
@@ -119,4 +331,358 @@ await File.WriteAllTextAsync(
...
+            JsonObject legacyState => legacyState,
+            _ => throw new InvalidDataException($"'{LegacyStateProperty}' must be a JSON object.")
+        };
+        var claimedSectionNames = migrationState[ClaimedSectionsProperty] switch
+        {
+            null => [],
+            JsonValue claimedSectionsValue => JsonSerializer.Deserialize<string[]>(
+                claimedSectionsValue.GetValue<string>()) ?? [],
+            _ => throw new InvalidDataException($"'{ClaimedSectionsProperty}' must be a JSON string.")
+        };
```
[the remark this answers is not in the document]
Comment (spboyer) [reply src/Aspire.Hosting/Pipelines/Internal/FileDeploymentStateManager.cs:760]: Fixed in 168629d95. Present claim metadata now rejects JSON `null`, null elements, and blank elements so malformed metadata follows the existing fail-closed path. Regression cases cover both serialized `null` and `[null]`.

File: src/Aspire.Hosting/Pipelines/Internal/FileDeploymentStateManager.cs [hunk 12]
```diff
@@ -96,20 +278,50 @@ protected override async Task SaveStateToStorageAsync(JsonObject state, Cancella
...
-                cancellationToken).ConfigureAwait(false);
+            using (await FileLock.AcquireAsync($"{deploymentStatePath}.lock", cancellationToken).ConfigureAwait(false))
+            {
+                if (sectionName is null)
+                {
+                    _currentState = state.DeepClone().AsObject();
+                    _legacyState = [];
+                    _legacyStateSnapshot = [];
+                    _claimedSectionNames.Clear();
+                    _legacyFallbackDisabled = true;
```
[the remark this answers is not in the document]
Comment (spboyer) [reply src/Aspire.Hosting/Pipelines/Internal/FileDeploymentStateManager.cs]: Fixed in 168629d95. Full saves now normalize through the same flatten/unflatten representation used by canonical persistence before updating authoritative migration state. The backward-compatibility test now verifies a restart reads the nested section correctly.

File: src/Aspire.Hosting/DistributedApplicationBuilder.cs [hunk 13]
```diff
@@ -304,9 +306,25 @@ public DistributedApplicationBuilder(DistributedApplicationOptions options)
...
+            var isSourceFileAppHost = !string.IsNullOrEmpty(appHostFilePath) &&
+                !string.Equals(Path.GetExtension(appHostFilePath), ".csproj", StringComparison.OrdinalIgnoreCase);
+            var appHostIdentityPath = isSourceFileAppHost
+                ? Path.GetFullPath(appHostFilePath!)
+                : AppHostPath;
+            var normalizedAppHostIdentityPath = isSourceFileAppHost && !OperatingSystem.IsWindows()
+                ? appHostIdentityPath
+                : appHostIdentityPath.ToLowerInvariant();
+            var appHostPathShaBytes = SHA256.HashData(Encoding.UTF8.GetBytes(normalizedAppHostIdentityPath));
             appHostPathSha = Convert.ToHexString(appHostPathShaBytes);
```
[the remark this answers is not in the document]
Comment (spboyer) [reply src/Aspire.Hosting/DistributedApplicationBuilder.cs]: Fixed in cd16d91eb. AppHost:PathSha256 remains the historical directory identity for browser profiles, Compose projects, and other persistent run-mode resources. Deployment state now uses the new source-specific AppHost:DeploymentStatePathSha256 key with an explicit legacy deployment-state fallback.

File: src/Aspire.Hosting/Pipelines/IDeploymentStateManager.cs [hunk 14]
```diff
@@ -25,6 +25,15 @@ public interface IDeploymentStateManager
...
     Task<DeploymentStateSection> AcquireSectionAsync(string sectionName, CancellationToken cancellationToken = default);
 
+    /// <summary>
+    /// Acquires a specific section from the current deployment state without falling back to legacy state.
+    /// </summary>
+    /// <param name="sectionName">The name of the section to acquire (e.g., "Parameters", "Azure").</param>
+    /// <param name="cancellationToken">The cancellation token.</param>
+    /// <returns>A StateSection containing the current section data and version information.</returns>
+    Task<DeploymentStateSection> AcquireCurrentSectionAsync(string sectionName, CancellationToken cancellationToken = default)
+        => AcquireSectionAsync(sectionName, cancellationToken);
```
[the remark this answers is not in the document]
Comment (spboyer) [reply src/Aspire.Hosting/Pipelines/IDeploymentStateManager.cs:39]: Kept the default intentionally to avoid a source-breaking change for external implementations of this public experimental interface. The contract now explicitly requires managers with legacy fallback to override it; FileDeploymentStateManager does so, while managers without fallback safely inherit AcquireSectionAsync. This preserves compatibility while retaining current-only semantics where they differ.

File: src/Aspire.Hosting/DistributedApplicationBuilder.cs [hunk 15]
```diff
@@ -299,15 +303,32 @@ public DistributedApplicationBuilder(DistributedApplicationOptions options)
...
+            // Source-file and polyglot AppHosts can share a host process and project directory,
+            // so use the actual source file to keep their deployment state isolated.
+            var isSourceFileAppHost = !string.IsNullOrEmpty(appHostFilePath) &&
+                !string.Equals(Path.GetExtension(appHostFilePath), ".csproj", StringComparison.OrdinalIgnoreCase);
+            var appHostIdentityPath = isSourceFileAppHost
+                ? Path.GetFullPath(appHostFilePath!)
+                : AppHostPath;
+            var normalizedAppHostIdentityPath = isSourceFileAppHost && !OperatingSystem.IsWindows()
+                ? appHostIdentityPath
+                : appHostIdentityPath.ToLowerInvariant();
```
[the remark this answers is not in the document]
Comment (spboyer) [reply src/Aspire.Hosting/DistributedApplicationBuilder.cs:325]: Fixed in 520067db7. Deployment identity now canonicalizes the actual filesystem entry casing on macOS and Windows: aliases on default case-insensitive APFS converge, while differently-cased files on a case-sensitive macOS volume remain distinct. The regression test uses a real temp file and branches on the containing filesystem behavior rather than the OS name.

File: src/Aspire.Hosting/Pipelines/Internal/FileDeploymentStateManager.cs [hunk 16]
```diff
@@ -119,4 +338,380 @@ await File.WriteAllTextAsync(
...
+            if (originalExists == savedExists && JsonNode.DeepEquals(originalValue, savedValue))
+            {
+                return (latestExists, latestValue?.DeepClone());
+            }
+
+            if (originalValue is JsonObject originalObject &&
+                savedValue is JsonObject savedObject)
+            {
+                return (true, ApplyChanges(latestValue, originalObject, savedObject));
+            }
```
[the remark this answers is not in the document]
Comment (spboyer) [reply src/Aspire.Hosting/Pipelines/Internal/FileDeploymentStateManager.cs]: Fixed in 28563b49f. When a saved object was absent from the writer’s original snapshot, the three-way merge now uses an empty object baseline and preserves independently added leaves from the latest state. Added a regression with stale writers adding Settings:Features:A and Settings:Features:B.

File: src/Aspire.Hosting/Pipelines/Internal/FileDeploymentStateManager.cs [hunk 17]
```diff
@@ -52,20 +64,125 @@ internal static bool IsValidEnvironmentName(string environmentName)
...
+        $"{canonicalStatePath}.migration";
+
+    internal static async Task<JsonObject> LoadEffectiveStateAsync(
+        string canonicalStatePath,
+        string? legacyStatePath,
+        CancellationToken cancellationToken = default)
+    {
+        JsonObject currentState;
+        MigrationState migrationState;
+        using (await FileLock.AcquireAsync($"{canonicalStatePath}.lock", cancellationToken).ConfigureAwait(false))
```
[the remark this answers is not in the document]
Comment (spboyer) [reply src/Aspire.Hosting/Pipelines/Internal/FileDeploymentStateManager.cs]: Fixed in 788cf54d2. Every canonical and legacy state lock is now acquired through a wrapper that applies `DirectoryHelper.CreateWithOwnerOnlyPermissions` before the lock file can create its parent. This covers synchronous and asynchronous effective-state loaders plus manager load, save, and clear paths. `LoadStateAsync_CreatesDirectory_WithUserOnlyPermissions` covers the fresh read path.

File: src/Aspire.Hosting/Pipelines/Internal/FileDeploymentStateManager.cs [hunk 18]
```diff
@@ -119,4 +338,380 @@ await File.WriteAllTextAsync(
...
+        };
+
+    /// <inheritdoc/>
+    protected override async Task ClearStateStorageAsync(CancellationToken cancellationToken)
+    {
+        _currentState = [];
+        _legacyState = [];
+        _legacyStateSnapshot = [];
+        _claimedSectionNames.Clear();
+        _legacyFallbackDisabled = true;
```
[the remark this answers is not in the document]
Comment (spboyer) [reply src/Aspire.Hosting/Pipelines/Internal/FileDeploymentStateManager.cs]: Fixed in 788cf54d2. Clear now stages an authoritative empty migration sidecar and publishes the empty in-memory view only after that sidecar commit succeeds. If lock acquisition or sidecar writing fails, the previous in-memory state remains intact; `CanceledClearPreservesCurrentState` covers the pre-commit cancellation case.

File: tests/Aspire.Hosting.Tests/DistributedApplicationBuilderTests.cs [hunk 19]
```diff
@@ -356,6 +359,372 @@ public void PathShaDiffersForDifferentPaths()
...
+
+            var options = new DistributedApplicationOptions
+            {
+                ProjectDirectory = projectDirectory.FullName,
+                ProjectName = projectName,
+                AppHostFilePath = Path.Combine(projectDirectory.FullName, "apphost.ts"),
+                Args = ["--publisher", "manifest"]
+            };
+
+            var builder = (DistributedApplicationBuilder)DistributedApplication.CreateBuilder(options);
```
[the remark this answers is not in the document]
Comment (spboyer) [reply tests/Aspire.Hosting.Tests/DistributedApplicationBuilderTests.cs:479]: Fixed in 82db9cdf1. The test now removes the canonical deployment directory it creates in addition to the legacy directory.

File: src/Aspire.Hosting/Pipelines/Internal/FileDeploymentStateManager.cs [hunk 20]
```diff
@@ -119,4 +338,430 @@ await File.WriteAllTextAsync(
...
+            claimedSectionNames.Add(path);
+        }
+    }
+
+    private static JsonNode? ApplyChanges(JsonNode? latestValue, JsonNode? originalValue, JsonNode? savedValue)
+    {
+        if (originalValue is JsonObject originalObject &&
+            savedValue is JsonObject savedObject)
+        {
+            var result = latestValue?.DeepClone() as JsonObject ?? [];
```
[the remark this answers is not in the document]
Comment (spboyer) [reply src/Aspire.Hosting/Pipelines/Internal/FileDeploymentStateManager.cs:556]: Fixed in 82db9cdf1. Scalar-wrapper and object structural transitions are now merged atomically, with a concurrent stale-writer regression preventing hybrid state.

File: src/Aspire.Hosting/Pipelines/Internal/FileDeploymentStateManager.cs [hunk 21]
```diff
@@ -52,20 +64,125 @@ internal static bool IsValidEnvironmentName(string environmentName)
...
+        }
+
+        JsonObject legacyState = [];
+        if (!string.IsNullOrEmpty(legacyStatePath))
+        {
+            using (await AcquireStateLockAsync(legacyStatePath, cancellationToken).ConfigureAwait(false))
+            {
+                legacyState = await LoadStateFileAsync(legacyStatePath, cancellationToken).ConfigureAwait(false);
+            }
+        }
```
[the remark this answers is not in the document]
Comment (spboyer) [reply src/Aspire.Hosting/Pipelines/Internal/FileDeploymentStateManager.cs:139]: Fixed in 82db9cdf1. Malformed or unreadable legacy fallback state is treated as unavailable without discarding canonical state across async, sync, and manager load paths.

File: src/Aspire.Hosting/Pipelines/Internal/FileDeploymentStateManager.cs [hunk 22]
```diff
@@ -52,20 +64,148 @@ internal static bool IsValidEnvironmentName(string environmentName)
...
+        CancellationToken cancellationToken = default)
+    {
+        JsonObject currentState;
+        MigrationState migrationState;
+        using (await AcquireStateLockAsync(canonicalStatePath, cancellationToken).ConfigureAwait(false))
+        {
+            currentState = await LoadStateFileAsync(canonicalStatePath, cancellationToken).ConfigureAwait(false);
+            try
+            {
+                migrationState = await LoadMigrationStateFileAsync(canonicalStatePath, cancellationToken).ConfigureAwait(false);
```
[the remark this answers is not in the document]
Comment (spboyer) [reply src/Aspire.Hosting/Pipelines/Internal/FileDeploymentStateManager.cs]: Fixed in c3c949d79. All three load paths now read valid migration metadata first and bypass canonical parsing when CurrentState is authoritative; malformed or absent sidecar metadata still falls back to the canonical file. The regression also verifies the recovered manager can save and survive another restart.

**Review by mitchdenny — APPROVED** [review]
Approving. Reviewed the deployment-state isolation change in isolation; the source-file identity + immutable legacy read-through migration is sound and well-tested (concurrency, tombstones, partial/repeated migration, malformed metadata).

Two edge cases were found and fixed on the branch during review, each with a regression test verified to fail without the fix:

1. `LoadDeploymentState` computed state paths outside its try/catch, so an environment name outside `[a-zA-Z0-9_-]` threw `ArgumentException` from builder construction in publish mode. Path resolution is now inside the guarded block.
2. Legacy state reads acquired the cross-process lock without a `File.Exists` guard, manufacturing phantom legacy directories/locks for brand-new source AppHosts and re-coupling siblings on a shared lock. Reads are now gated on legacy-file existence (safe, since legacy state is immutable in this model).

Also reconciled the `SelfUpdateChannelPersistenceTests` disable with #19733 (now merged), and merged main. CI is fully green (385 pass / 0 fail).

Follow-up filed: #19734 to update the deployment-state-caching docs for the new source-AppHost layout.
